### DISCUSSION TITLE
OpenMLDB Error System

### DISCUSSION THREAD

**Discussion by jingchen2222** [4paradigm/OpenMLDB#429] [Ideas]
# Motivation
Users can get frustrated when they use OpenMLDB in the wrong way. e.g. SQL Syntax errors, unsupported data types, unsupported query on specific running mode. What's worse, the error message can be misleading in some cases which might make users more confused.
So it's very important to improve, refactor if need, the whole error message system in a way to make the error message clear and readable.

- [x] ERROR CODE to identify the error types.
- [ ] ERROR message had better follow some pattern.
- [x] TRACE message can be Exposed when the DB turn on `TRACE_ENABLE` flag
- [ ] Survey Error System in Popular DB: MySQL/porstgre/Sqlite/spark sql/redis/Tidb @aceforeverd @jingchen2222 
- [ ] Design Error Message Pattern
- [ ]

**Comment by jingchen2222** [comment]
# Error Code

## Success Code
    kOk = 0;
    kRunning = 1;

## General error
    kRunError = 2;
    kTypeError = 3;
    kNullPointer = 10;
    kNullInputPointer = 11;
    kNullOutputPointer = 12;

## Unsupport
    kUnSupport = 100;
    kUnsupportSql = 101;
    kUnsupportPlan = 102;
    kUnsupportPerformanceSensitive = 103;
    kUnsupportDistribution = 104;
    kUnSupportCommand = 105;

## PlanError
    kPlanError = 1000;
    kSyntaxError = 1001;
    kSqlAstError = 1002;
    kLogicalPlanError = 1003;
    kPhysicalPlanError = 1004;
    kExecutionPlanError = 1005;
    kEngineModeError = 1006;
    kEngineCacheError = 1007;

## CodegenError
    kCodegenError = 1100;
    kCodegenLoadValueError = 1101;
    kCodegenStoreValueError = 1102;
    kCodegenExpressionError = 1103;
    kCodegenCallFunctionError = 1104;
    kCodegenUdfError = 1105;
    kCodegenUdafError = 1106;
    kCodegenGetFieldError = 1107;
    kCodegenEncodeError = 1108;
    kCodegenDecodeError = 1109;
## CatalogError
    kCatalogError = 1200;
    kNoDatabase = 1201;
    kTableNotFound = 1202;
    kProcedureNotFound = 1203;
    kColumnNotFound = 1204;
    kIndexNotFound=1205;
    kTableExists = 1206;

## CodecError
    kCodecError = 1300;
    kSchemaCodecError = 1301;

## Runtime Error
    kRuntimeError = 1400;
    kJitError = 1401;
    kCoreAPIRuntimeError = 1402;

## RPC Error
    kRpcError = 1500;
    kConnError = 1501;
    kRequestError = 1502;
    kResponseError = 1503;
    kTimeoutError = 1504;
    kCallRpcMethodError = 1505;

## Storage Error
    kStorageError = 1600;
    kTablePutFailed = 1601;

## External Error
    kExternalError = 1700;
    kFileIOError = 1701;

## Test Error
    kTestError = 1800;
    kTestEngineError = 1801;
    kSqlCaseError = 1802;
    kTestRunSessionError = 1803;

**Comment by jingchen2222** [comment]
TODO: Survey Error System in Popular DB: MySQL/porstgre/Sqlite/spark sql/redis/Tidb/Voltdb @aceforeverd @jingchen2222

**Reply by aceforeverd** [reply]
take 2, 3 first

**Comment by jingchen2222** [comment]
# Survey Error System in Popular DB

**Comment by aceforeverd** [comment]
## Postgres: 

Three message type:
1. primary: shot & factual, no impl detail
2. detail
3. hint

Formating/Writing guide:
- lower case for message wording, upper case for SQL commands and key words
- past tense if an attempt to do something failed, but could perhaps succeed next time; present tense if the failure is certainly permanent.

links
- https://www.postgresql.org/docs/9.1/error-style-guide.html
- https://www.postgresql.org/docs/9.1/nls-programmer.html#NLS-GUIDELINES

**Comment by jingchen2222** [comment]
## MySQL
### Error reference
- Error number
- Error symbol
- Message pattern
https://dev.mysql.com/doc/mysql-errors/5.7/en/server-error-reference.html
We can learn the pattern from mysql.

https://dev.mysql.com/doc/refman/5.7/en/perror.html
> Command-Line Error Message Interface
The perror program provides information from the command line about error numbers. See Section 4.8.2, “perror — Display MySQL Error Message Information”.

```
$> perror 1231
MySQL error code 1231 (ER_WRONG_VALUE_FOR_VAR): Variable '%-.64s' can't
be set to the value of '%-.200s'
```

### Error message format
https://github.com/mysql/mysql-server/blob/3290a66c89eb1625a7058e0ef732432b6952b435/libmysql/errmsg.cc

https://sourcegraph.com/github.com/jimdb-org/jimdb/-/blob/jimsql/jim-sdk/jim-common/src/main/java/io/jimdb/common/exception/ErrorCode.java

https://github.com/jimdb-org/jimdb/blob/927e4447879189597dbe7f91a7fe0e8865107ef4/jimsql/jim-sdk/jim-core/src/main/java/io/jimdb/core/expression/ColumnExpr.java
```java
public abstract class BaseException extends RuntimeException {
//...
 public static final String message(ErrorCode code, String... params) {
    String msg = (CODE_MESSAGES.isEmpty() || !CODE_MESSAGES.containsKey(code)) ? code.getMessage() : CODE_MESSAGES.get(code);
    if (params != null) {
      msg = String.format(msg, params);
    }
    return msg;
  }
}
```
```java
public final class DBException extends BaseException {
  private DBException(ErrorModule module, ErrorCode code, String message, Throwable cause) {
    super(module, code, message, cause);
  }

  public static DBException get(ErrorModule module, ErrorCode code, String... params) {
    return get(module, code, true, params);
  }

  public static DBException get(ErrorModule module, ErrorCode code, boolean format, String... params) {
    return new DBException(module, code, format ? message(code, params) : params[0], null);
  }

  public static DBException get(ErrorModule module, ErrorCode code, Throwable cause, String... params) {
    return new DBException(module, code, message(code, params), cause);
  }
}
```

```java
// ER_BAD_FIELD_ERROR(1054, "42S22", "Unknown column '%s' in '%s'"),
if (result.offset == -1) {
      throw DBException.get(ErrorModule.PARSER, ErrorCode.ER_BAD_FIELD_ERROR, oriCol, oriTable);
    }
```

**Comment by jingchen2222** [comment]
## TiDB
> TiDB is compatible with the error codes in MySQL, and in most cases returns the same error code as MySQL. For a list of error codes for MySQL, see MySQL 5.7 Error Message Reference. In addition, TiDB has the following unique error codes:
https://docs.pingcap.com/tidb/stable/error-codes#error-codes-and-troubleshooting
